integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+	lcConf.Auth.Enabled = true
+	lcConf.Auth.Preference.SetSecondFactor("off")
+	lcConf.Proxy.Enabled = true
+	lcConf.Proxy.DisableWebService = false
+	lcConf.Proxy.DisableWebInterface = true
+	lcConf.SSH.Enabled = false
+	lcConf.Apps.Enabled = false
+
+	err = p.leafCluster.CreateEx(p.rootCluster.Secrets.AsSlice(), lcConf)
+	require.NoError(t, err)
+	err = p.rootCluster.CreateEx(p.leafCluster.Secrets.AsSlice(), rcConf)
+	require.NoError(t, err)
+
+	err = p.leafCluster.Start()
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		p.leafCluster.StopAll()
+	})
+	err = p.rootCluster.Start()
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		p.rootCluster.StopAll()
+	})
+
+	raConf := service.MakeDefaultConfig()
+	raConf.DataDir, err = ioutil.TempDir("", "app-server-"+p.rootCluster.Secrets.SiteName)
+	require.NoError(t, err)
+	raConf.Token = "static-token-value"
+	raConf.AuthServers = []utils.NetAddr{
+		utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        net.JoinHostPort(Loopback, p.rootCluster.GetPortWeb()),
+		},
+	}
+	raConf.Auth.Enabled = false
+	raConf.Proxy.Enabled = false
+	raConf.SSH.Enabled = false
+	raConf.Apps.Enabled = true
+	raConf.Apps.Apps = []service.App{
+		service.App{
+			Name:       p.rootAppName,
+			URI:        rootServer.URL,
+			PublicAddr: p.rootAppPublicAddr,
+		},
+	}
+	p.rootAppServer, err = p.rootCluster.StartApp(raConf)
+	require.NoError(t, err)
+
+	laConf := service.MakeDefaultConfig()
+	laConf.DataDir, err = ioutil.TempDir("", "app-server-"+p.leafCluster.Secrets.SiteName)
+	require.NoError(t, err)
+	laConf.Token = "static-token-value"
+	laConf.AuthServers = []utils.NetAddr{
+		utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        net.JoinHostPort(Loopback, p.leafCluster.GetPortWeb()),
+		},
+	}
+	laConf.Auth.Enabled = false
+	laConf.Proxy.Enabled = false
+	laConf.SSH.Enabled = false
+	laConf.Apps.Enabled = true
+	laConf.Apps.Apps = []service.App{
+		service.App{
+			Name:       p.leafAppName,
+			URI:        leafServer.URL,
+			PublicAddr: p.leafAppPublicAddr,
+		},
+	}
+	p.leafAppServer, err = p.leafCluster.StartApp(laConf)
+	require.NoError(t, err)
+
+	return p
+}
+
+// createUser are user within the root cluster.
+func (p *pack) createUser(t *testing.T) services.User {
+	p.username = uuid.New()
+	p.password = uuid.New()
+
+	user, err := services.NewUser(p.username)
+	require.NoError(t, err)
+
+	role := services.RoleForUser(user)
+	role.SetLogins(services.Allow, []string{p.username})
+	err = p.rootCluster.Process.GetAuthServer().UpsertRole(context.Background(), role)
+	require.NoError(t, err)
+
+	user.AddRole(role.GetName())
+	err = p.rootCluster.Process.GetAuthServer().CreateUser(context.Background(), user)
+	require.NoError(t, err)
+
+	err = p.rootCluster.Process.GetAuthServer().UpsertPassword(user.GetName(), []byte(p.password))
+	require.NoError(t, err)
+
+	return user
+}
+
+// createWebSession creates a Web UI session within the root cluster.
+func (p pack) createWebSession(t *testing.T) (string, string) {
```

Comment:
I would use named return args here so it's clearer for the reader what is being returned.